cluster-autoscaler/simulator/clustersnapshot/predicate/predicate_snapshot_test.go

```diff
@@ -161,58 +355,860 @@ func validTestCases(t *testing.T) []modificationTestCase {
+				// The pod-owned claim should get added to the DRA snapshot, with a reservation for the pod.
+				// LocalResourceSlices from the NodeInfo should get added to the DRA snapshot.
+				modifiedState: snapshotState{
+					nodes:      []*apiv1.Node{node},
+					podsByNode: map[string][]*apiv1.Pod{node.Name: {podWithClaims}},
+					draSnapshot: drasnapshot.NewSnapshot(
+						map[drasnapshot.ResourceClaimId]*resourceapi.ResourceClaim{
+							drasnapshot.GetClaimId(podOwnedClaim): drautils.TestClaimWithPodReservations(drautils.TestClaimWithAllocation(podOwnedClaim, podOwnedClaimAlloc), podWithClaims),
+							drasnapshot.GetClaimId(sharedClaim):   drautils.TestClaimWithPodReservations(drautils.TestClaimWithAllocation(sharedClaim, sharedClaimAlloc), podWithClaims),
+						},
+						map[string][]*resourceapi.ResourceSlice{node.Name: resourceSlices}, nil, deviceClasses),
+				},
+			},
+			{
+				name: "adding LocalResourceSlices for an already tracked Node is an error",
+				state: snapshotState{
+					draSnapshot: drasnapshot.NewSnapshot(nil, map[string][]*resourceapi.ResourceSlice{node.Name: resourceSlices}, nil, deviceClasses),
+				},
+				op: func(snapshot clustersnapshot.ClusterSnapshot) error {
+					return snapshot.AddNodeInfo(framework.NewNodeInfo(node, resourceSlices))
+				},
+				// LocalResourceSlices for the Node already exist in the DRA snapshot, so trying to add them again should be an error.
+				wantErr: cmpopts.AnyError,
+				// The state shouldn't change on error.
+				modifiedState: snapshotState{
+					draSnapshot: drasnapshot.NewSnapshot(nil, map[string][]*resourceapi.ResourceSlice{node.Name: resourceSlices}, nil, deviceClasses),
+				},
+			},
+			{
+				name: "adding already tracked pod-owned ResourceClaims is an error",
+				state: snapshotState{
+					draSnapshot: drasnapshot.NewSnapshot(
+						map[drasnapshot.ResourceClaimId]*resourceapi.ResourceClaim{
+							drasnapshot.GetClaimId(sharedClaim):   drautils.TestClaimWithAllocation(sharedClaim, sharedClaimAlloc),
+							drasnapshot.GetClaimId(podOwnedClaim): podOwnedClaim.DeepCopy(),
+						}, nil, nil, deviceClasses),
+				},
+				op: func(snapshot clustersnapshot.ClusterSnapshot) error {
+					podInfo := framework.NewPodInfo(podWithClaims, []*resourceapi.ResourceClaim{
+						drautils.TestClaimWithAllocation(podOwnedClaim, podOwnedClaimAlloc),
+						drautils.TestClaimWithAllocation(sharedClaim, sharedClaimAlloc),
+					})
+					return snapshot.AddNodeInfo(framework.NewNodeInfo(node, resourceSlices, podInfo))
+				},
+				// The pod-owned claim already exists in the DRA snapshot, so trying to add it again should be an error.
+				wantErr: cmpopts.AnyError,
+				// The state shouldn't change on error.
+				// TODO(DRA): Until transaction-like clean-up is implemented in AddNodeInfo, the state is not cleaned up on error. Make modifiedState identical to initial state after the clean-up is implemented.
```

Comment:
Is lack of "transactions" not a blocker for MVP? I am a bit worried that this might easily result in a broken state. Could this be straightforward to implement by adding defer RemoveNodeInfo?